src/rsa/signing.rs

```diff
@@ -125,11 +125,34 @@ impl RSAKeyPair {
                 if p.bit_length() != q.bit_length() {
                     return Err(error::Unspecified);
                 }
+                let q = try!(q.into_odd_positive());
+                let n = try!(n.into_modulus::<N>());
+
+                let q_mod_n_decoded = {
+                    let q = try!(q.try_clone());
+                    try!(q.into_elem_decoded(&n))
+                };
+
                 // XXX: |p < q| is actual OK, it seems, but our implementation
                 // of CRT-based moduluar exponentiation used requires that
                 // |q > p|. (|p == q| is just wrong.)
-                let q = try!(q.into_odd_positive());
                 try!(bigint::verify_less_than(&q, &p));
+                {
+                    let p_mod_n = {
+                        let p = try!(p.try_clone());
+                        try!(p.into_elem_decoded(&n))
+                    };
+                    let p_minus_q = {
+                        let p_minus_q = try!(bigint::elem_sub(
+                                p_mod_n, &q_mod_n_decoded, &n));
+                        try!(p_minus_q.into_positive())
+                    };
+                    let min_pq_bitlen_diff = try!(half_n_bits.try_sub(
+                            bits::BitLength::from_usize_bits(100)));
+                    if p_minus_q.bit_length() < min_pq_bitlen_diff {
+                        return Err(error::Unspecified);
+                    }
```

Comment:
Ok. I didn't the exact translation from numbers to bits through very carefully the first time around because it didn't appear to me to matter in practice, so I'm +1 to just adding a comment.